log4j2-ecs-layout/src/main/java/co/elastic/logging/log4j2/MdcSerializer.java

```diff
@@ -24,27 +24,42 @@
  */
 package co.elastic.logging.log4j2;
 
+import java.util.ServiceLoader;
+
 import co.elastic.logging.EcsJsonSerializer;
 import co.elastic.logging.JsonUtils;
 import org.apache.logging.log4j.core.LogEvent;
 import org.apache.logging.log4j.util.TriConsumer;
 
-interface MdcSerializer {
+public interface MdcSerializer {
 
     void serializeMdc(LogEvent event, StringBuilder builder);
 
     class Resolver {
 
         public static MdcSerializer resolve() {
+            MdcSerializer serializer = resolveUserDefinedMdcSerializer();
+            if (serializer != null) {
+                return serializer;
+            }
             try {
                 LogEvent.class.getMethod("getContextData");
                 return (MdcSerializer) Class.forName("co.elastic.logging.log4j2.MdcSerializer$UsingContextData").getEnumConstants()[0];
-            } catch (Exception ignore) {
-            } catch (LinkageError ignore) {
+            } catch (Exception | LinkageError ignore) {
             }
             return UsingContextMap.INSTANCE;
         }
 
+        private static MdcSerializer resolveUserDefinedMdcSerializer() {
```

Comment:
I don't think using `ServiceLoader` is a good idea here. This changes the behaviour of the logging based just on the classpath contents. E.g. if a user pulls in a library which accidentally includes a `MdcSerializer`, that one will be picked up and suddenly change the logging as a side effect.

I'd like to have this explicitly configurable on the `EcsLayout` instead:
```
<EcsLayout serviceName="my-app" ... mdcSerializer="foo.bar.CustomSerializerImpl"/>
```

The `mdcSerializer` config option takes the FQN of the `MdcSerializer` to use. An empty string should be the default, causing the existing `DefaultMdcSerializer` to be used. The serializer instance is then created simply via reflection instead of using the `ServiceLoader`. We should clarify on the `MdcSerializer` javadoc that implementations must be public and have a public zero-args constructor.

Reply:
I also prefer this approach. I was just following the existing style for `ObjectMessageJacksonSerializer`. It seemed logical to load them in a consistent way.